### PR TITLE
DOPE-218: localized parameter and iterator manager creation

### PR DESCRIPTION
**Pull request by martinagallati-ergon**

### CODE REVIEW

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/Clause.kt [hunk 1]
```diff
@@ -1,14 +1,13 @@
...
 import ch.ergon.dope.resolvable.Resolvable
-import ch.ergon.dope.resolvable.expression.unaliased.type.ParameterManager
-import ch.ergon.dope.resolvable.expression.unaliased.type.collection.IteratorManager
 
 interface Clause : Resolvable {
     fun build(): DopeQuery {
-        ParameterManager.resetCounter()
-        IteratorManager.resetCounter()
-        return toDopeQuery()
+        // ParameterManager.resetCounter()
```
Comment (jansigi) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/Clause.kt:9]: Why is this commented? Remove it.

File: core/src/main/kotlin/ch/ergon/dope/DopeQuery.kt [hunk 2]
```diff
@@ -1,3 +1,11 @@
 package ch.ergon.dope
 
-data class DopeQuery(val queryString: String, val parameters: Map<String, Any>)
+import ch.ergon.dope.resolvable.expression.unaliased.type.ParameterManager
+import ch.ergon.dope.resolvable.expression.unaliased.type.collection.IteratorManager
+
+data class DopeQuery(val queryString: String, val parameters: Map<String, Any>, val manager: DopeQueryManager)
```
Comment (jansigi) [inline core/src/main/kotlin/ch/ergon/dope/DopeQuery.kt:6]: Why do we have to store the manager in the DopeQuery? According to my understanding, the manager is only needed in the `.toDopeType(manager)` function?

File: crystal-map-connector/src/test/kotlin/ch/ergon/dope/extensions/aggregator/ArrayAggTest.kt [hunk 3]
```diff
@@ -10,10 +11,18 @@ import ch.ergon.dope.helper.someCMStringList
 import ch.ergon.dope.resolvable.expression.unaliased.aggregator.AggregateQuantifier.ALL
 import ch.ergon.dope.resolvable.expression.unaliased.aggregator.ArrayAggregateExpression
 import ch.ergon.dope.toDopeType
+import kotlin.test.BeforeTest
 import kotlin.test.Test
 import kotlin.test.assertEquals
 
 class ArrayAggTest {
+    private lateinit var manager: DopeQueryManager
```
Comment (jansigi) [inline crystal-map-connector/src/test/kotlin/ch/ergon/dope/extensions/aggregator/ArrayAggTest.kt:19]: Maybe also add a `ManagerDependentTest` that has the `setUp` function

**Review by jansigi — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/Clause.kt (hunk 1 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/Clause.kt:9]: done

File: crystal-map-connector/src/test/kotlin/ch/ergon/dope/extensions/aggregator/ArrayAggTest.kt (hunk 3 above)
Comment (martinagallati-ergon) [reply crystal-map-connector/src/test/kotlin/ch/ergon/dope/extensions/aggregator/ArrayAggTest.kt:19]: done

**Review by jansigi — APPROVED** [review]

File: core/src/main/kotlin/ch/ergon/dope/DopeQuery.kt [hunk 4]
```diff
@@ -1,3 +1,11 @@
 package ch.ergon.dope
 
+import ch.ergon.dope.resolvable.expression.unaliased.type.ParameterManager
+import ch.ergon.dope.resolvable.expression.unaliased.type.collection.IteratorManager
+
 data class DopeQuery(val queryString: String, val parameters: Map<String, Any>)
+
+class DopeQueryManager {
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/DopeQuery.kt:8]: maybe move it into an own class - It has not directly something to do with DopeQuery

Comment (pgruntz) [reply core/src/main/kotlin/ch/ergon/dope/DopeQuery.kt:8]: And maybe move the ParameterManager and the IteratorManager in that file as well (or at least in the same package)

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/Clause.kt [hunk 5]
```diff
@@ -1,14 +1,11 @@
...
 import ch.ergon.dope.resolvable.Resolvable
-import ch.ergon.dope.resolvable.expression.unaliased.type.ParameterManager
-import ch.ergon.dope.resolvable.expression.unaliased.type.collection.IteratorManager
 
 interface Clause : Resolvable {
     fun build(): DopeQuery {
-        ParameterManager.resetCounter()
-        IteratorManager.resetCounter()
-        return toDopeQuery()
+        return toDopeQuery(manager = DopeQueryManager())
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/Clause.kt:9]: use an expression body here

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt [hunk 6]
```diff
@@ -1,14 +1,15 @@
...
 import ch.ergon.dope.validtype.ArrayType
 import ch.ergon.dope.validtype.BooleanType
 import ch.ergon.dope.validtype.NumberType
 import ch.ergon.dope.validtype.StringType
 import ch.ergon.dope.validtype.ValidType
 
-object ParameterManager {
+class ParameterManager {
     var count: Int = 1
         get() = field++
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt:15]: please remove the _resetCounter()_-method, I think it's not needed anymore

Comment (pgruntz) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt:15]: and please remove it in the other manager as well

**Review by pgruntz — APPROVED** [review]

File: core/src/main/kotlin/ch/ergon/dope/DopeQuery.kt (hunk 4 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/DopeQuery.kt:8]: I moved all classes into one file named `Manager`.

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/Clause.kt (hunk 5 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/Clause.kt:9]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt (hunk 6 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Parameter.kt]: done. I also made the setter private that the user can't set the counter

**Pull request merged**